app/validators/file_type_allowed_validator.rb

```diff
@@ -0,0 +1,15 @@
+class FileTypeAllowedValidator < ActiveModel::EachValidator
+  # The list of allowed types is security sensitive because it may allow files with script
+  # capabilities to be uploaded leading to XSS attacks. Please be sure to know what you are
+  # doing when modifying this list.
+  VALID_FILE_TYPES = [".jpg", ".jpeg", ".pdf", ".png", ".heic", ".bmp", ".txt", ".tiff", ".gif"]
+
+  def validate_each(record, attr_name, value)
+    return if !value
+
+    extension = File.extname(value.tempfile.path)
```

Comment:
i concluded that this was the best way to get the extension by studying the test file uploads and also putting a breakpoint in and testing locally to see what `value` is here. it appears to be something of type `ActionDispatch::Http::UploadedFile` not a document, so our original tactic of getting `.upload.attachment.filename.extension` didn't work. there could be a better way to do this!

Reply:
We may be able to use `value.content_type` or `value.original_filename` as well. You can also just call `value.path` I think. https://api.rubyonrails.org/v6.0.3.2/classes/ActionDispatch/Http/UploadedFile.html

Reply:
jpeg is also a valid extension for jpg